### PR TITLE
Fix: "Code formatting button does not escape backticks"

### PR DESCRIPTION
**Pull request by yaya-usman**
Fixes:  [vector-im/element-web#21437](https://github.com/vector-im/element-web/issues/21437)

Issue reported by @robintown 

Signed-off-by: Yaya Usman [yaya-usman@users.noreply.github.com](mailto:yaya-usman@users.noreply.github.com) 

**Before:**
beforefix1

beforefix2

**After:** 
afterfix1
afterfix2

Type: defect

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix: "Code formatting button does not escape backticks" ([\#8181](https://github.com/matrix-org/matrix-react-sdk/pull/8181)). Contributed by @yaya-usman.

Preview: https://pr8181--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by yaya-usman** [comment]
@t3chguy can you help review this?

File: src/components/views/dialogs/ShareDialog.tsx [hunk 1]
```diff
@@ -155,7 +155,7 @@ export default class ShareDialog extends React.PureComponent<IProps, IState> {
             checkbox = <div>
                 <StyledCheckbox
                     checked={this.state.linkSpecificEvent}
-                    onClick={this.onLinkSpecificEventCheckboxClick}
+                    onChange={this.onLinkSpecificEventCheckboxClick}
```
Comment (t3chguy) [inline src/components/views/dialogs/ShareDialog.tsx:158]: This change is unrelated to what you are fixing here, please don't conflate your PRs

File: src/editor/operations.ts [hunk 2]
```diff
@@ -205,8 +206,13 @@ export function formatRangeAsCode(range: Range): void {
             parts.push(partCreator.newline());
         }
     } else {
-        toggleInlineFormat(range, "`");
-        return;
+        if (hasBacktick) {
+            parts.unshift(partCreator.plain("``"));
+            parts.push(partCreator.plain("``"));
```
Comment (t3chguy) [inline src/editor/operations.ts:211]: I think this needs to be a bit smarter, someone could have selected ```foo``bar``` and this would not work for that, also please write a test to exercise this behaviour to prevent it regressing for the future

Comment (robintown) [reply src/editor/operations.ts:211]: You can reference https://github.com/matrix-org/matrix-react-sdk/blob/afa60acbf123e6e739bcd6fe409485fd338c8295/src/editor/deserialize.ts#L203 for how this is done for message edits

Comment (t3chguy) [reply src/editor/operations.ts:211]: Also as per the current test failure, it seems to have broken the toggling behaviour of this feature, it needs to be smarter in that it looks at the number of backticks at the edge of the selection to know whether to add or remove backticks, and count the backticks within the selection (not at the end) to know how many to wrap in

File: src/components/views/dialogs/ShareDialog.tsx (hunk 1 above)
Comment (yaya-usman) [reply src/components/views/dialogs/ShareDialog.tsx:158]: i don't know why the commit is still showing even though i opened a new branch with this new PR

Comment (t3chguy) [reply src/components/views/dialogs/ShareDialog.tsx:158]: You probably branched straight from your other branch, instead of `develop` - easy way around that in future is to always switch back to develop first.

File: src/editor/operations.ts (hunk 2 above)
Comment (yaya-usman) [reply src/editor/operations.ts:211]: > I think this needs to be a bit smarter, someone could have selected `` foo``bar `` and this would not work for that (it'd write `` ``foo``bar`` ``, also please write a test to exercise this behaviour to prevent it regressing for the future

I am not really good with tests, hopefully i'll use your help to get it done.

File: src/components/views/dialogs/ShareDialog.tsx (hunk 1 above)
Comment (yaya-usman) [reply src/components/views/dialogs/ShareDialog.tsx:158]: > You probably branched straight from your other branch, instead of `develop` - easy way around that in future is to always switch back to develop first.

Noted, Thanks, i am learning a lot from you : )

**Comment by yaya-usman** [comment]
@t3chguy @robintown Hi, can you please review my PR  : ), also thank you guys for the tips.

**Review by t3chguy — CHANGES_REQUESTED** [review]
> @t3chguy @robintown Hi, can you please review my PR : )

You still have the unrelated change in it.

You should use existing utility functions like `longestBacktickSequence` and you should write a test to verify this does the right thing and doesn't break in the future

Thanks

**Comment by yaya-usman** [comment]
> > @t3chguy @robintown Hi, can you please review my PR : )
> 
> You still have the unrelated change in it.
> 
> You should use existing utility functions like `longestBacktickSequence` and you should write a test to verify this does the right thing and doesn't break in the future
> 
> Thanks

Ok, I will have to force-push to discard the unrelated change, as for test, I will try since I'm not good with unit testing.

**Comment by yaya-usman** [comment]
@t3chguy hi, here i am again : ), i have done all the things you requested be done.

File: test/editor/operations-test.js [hunk 3]
```diff
@@ -44,6 +45,23 @@ describe('editor/operations: formatting operations', () => {
...
+
+            const range = model.startRange(model.positionForOffset(0, false),
+                model.positionForOffset(13, false));  // hello ` world
+
+            expect(range.parts[0].text.trim().includes("`")).toBeTruthy();
+            expect(longestBacktickSequence(range.parts[0].text.trim())).toBe(1);
+            expect(model.serializeParts()).toEqual([{ "text": "hello ` world!", "type": "plain" }]);
+            toggleInlineFormat(range, "`".repeat(longestBacktickSequence(range.parts[0].text.trim()) + 1));
+            expect(model.serializeParts()).toEqual([{ "text": "``hello ` world``!", "type": "plain" }]);
+        });
```
Comment (t3chguy) [inline test/editor/operations-test.js:63]: This test does not call upon the method where your code lives `formatRangeAsCode`

File: src/editor/operations.ts [hunk 4]
```diff
@@ -205,8 +208,13 @@ export function formatRangeAsCode(range: Range): void {
...
     } else {
-        toggleInlineFormat(range, "`");
-        return;
+        if (hasBacktick) {
+            parts.unshift(partCreator.plain("`".repeat(longestBacktickSequence(range.text)+ 1)));
+            parts.push(partCreator.plain("`".repeat(longestBacktickSequence(range.text)+ 1)));
+        } else {
+            toggleInlineFormat(range, "`");
+            return;
+        }
```
Comment (t3chguy) [inline src/editor/operations.ts:217]: This is a great start but doesn't work for toggling. You should probably do something like

`toggleInlineFormat(range, "`".repeat(longestBacktickSequence(range.text)+ 1)`

Currently trying to undo the code by clicking the button a 2nd time just deletes the entire text.

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: src/editor/operations.ts (hunk 4 above)
Comment (yaya-usman) [reply src/editor/operations.ts]: @t3chguy hi, how about now, below is attached video of it being tested, i will be waiting for your review :

ezgif com-gif-maker
)

File: src/editor/operations.ts [hunk 5]
```diff
@@ -181,12 +182,15 @@ export function formatRangeAsCode(range: Range): void {
 
     const hasBlockFormatting = (range.length > 0)
         && range.text.startsWith("```")
-        && range.text.endsWith("```");
+        && range.text.endsWith("```")
+        && range.text.includes('\n');
+    const hasBacktick = (range.text.includes("`"))
+        && !range.text.startsWith("`")
+        && !range.text.endsWith("`");
```
Comment (t3chguy) [inline src/editor/operations.ts:189]: I think this variable name should be improved

File: src/editor/operations.ts [hunk 6]
```diff
@@ -205,8 +209,18 @@ export function formatRangeAsCode(range: Range): void {
...
         }
     } else {
-        toggleInlineFormat(range, "`");
-        return;
+        const fenceLen = longestBacktickSequence(range.text);
+        if (hasBacktick) {
+            parts.unshift(partCreator.plain("`".repeat(fenceLen + 1)));
+            parts.push(partCreator.plain("`".repeat(fenceLen+ 1)));
+        } else {
+            if (fenceLen === 0) {
```
Comment (t3chguy) [inline src/editor/operations.ts:217]: Can we collapse the else-ifs?

File: src/editor/operations.ts [hunk 7]
```diff
@@ -205,8 +209,18 @@ export function formatRangeAsCode(range: Range): void {
...
+                toggleInlineFormat(range, "`");
+                return;
+            }
+            toggleInlineFormat(range, "`".repeat(fenceLen));
+            return;
+        }
```
Comment (t3chguy) [inline src/editor/operations.ts:223]: This works but I think can be simplified, can you not call `toggleInlineFormat` in all branches here?

File: test/editor/operations-test.js [hunk 8]
```diff
@@ -44,6 +46,23 @@ describe('editor/operations: formatting operations', () => {
...
+
+            const range = model.startRange(model.positionForOffset(0, false),
+                model.positionForOffset(13, false));  // hello ` world
+
+            expect(range.parts[0].text.trim().includes("`")).toBeTruthy();
+            expect(longestBacktickSequence(range.parts[0].text.trim())).toBe(1);
+            expect(model.serializeParts()).toEqual([{ "text": "hello ` world!", "type": "plain" }]);
+            formatRangeAsCode(range);
+            expect(model.serializeParts()).toEqual([{ "text": "``hello ` world``!", "type": "plain" }]);
+        });
```
Comment (t3chguy) [inline test/editor/operations-test.js:64]: You should test that it works for longer sequences and that toggling it also works

**Review by t3chguy — COMMENTED** [review]
Good progress, it seems to work as expected now!

File: src/editor/operations.ts (hunk 5 above)
Comment (yaya-usman) [reply src/editor/operations.ts:189]: i followed same naming approach as the code above it which says, `hasBlockFormatting`, if you have a better name, kindly suggest one.

File: src/editor/operations.ts (hunk 6 above)
Comment (yaya-usman) [reply src/editor/operations.ts:217]: Ikr, Trust me, i tried many variations but it doesn't seem to work except i do it this way :(

File: test/editor/operations-test.js (hunk 8 above)
Comment (yaya-usman) [reply test/editor/operations-test.js:64]: > You should test that it works for longer sequences and that toggling it also works

I didn't wanted to write too much tests, it actually works for all the case you stated even for longer sequence, i attached a video up there, you probably missed it. I can drop new video for you to see

File: src/editor/operations.ts (hunk 7 above)
Comment (yaya-usman) [reply src/editor/operations.ts:223]: > This works but I think can be simplified, can you not call `toggleInlineFormat` in all branches here?

Unfortunately, that's the only way i could come up with, as stated earlier, i tried to simplify further but it doesn't work which prompted me to leave it this way

File: test/editor/operations-test.js (hunk 8 above)
Comment (t3chguy) [reply test/editor/operations-test.js:64]: There is no such thing as "too much tests" - tests protect against this feature being broken by accident by contributions later down the line, the video only tests it working at the time of this PR being made. Please write tests which exercise all edges. Its great work, just needs testing so it doesn't accidentally regress through the mistakes of others.

Comment (yaya-usman) [reply test/editor/operations-test.js:64]: > There is no such thing as "too much tests" - tests protect against this feature being broken by accident by contributions later down the line, the video only tests it working at the time of this PR being made. Please write tests which exercise all edges. Its great work, just needs testing so it doesn't accidentally regress through the mistakes of others.

Ok thanks but Can you pls be clear about the longer sequence you meant like give an example so i can write proper test for it.  if you mean escaping longer backticks in between text like hello```world, then i can just write test for that.

Comment (yaya-usman) [reply test/editor/operations-test.js]: I think i now understand you, I have written more tests with all the edge cases i could think of, i'll appreciate if there is more you think i should add

> > There is no such thing as "too much tests" - tests protect against this feature being broken by accident by contributions later down the line, the video only tests it working at the time of this PR being made. Please write tests which exercise all edges. Its great work, just needs testing so it doesn't accidentally regress through the mistakes of others.
> 
> Ok thanks but Can you pls be clear about the longer sequence you meant like give an example so i can write proper test for it. if you mean escaping longer backticks in between text like hello```world, then i can just write test for that.

I think i now understand you, I have written more tests with all the edge cases i could think of, i'll appreciate if there is more you think i should add

File: src/editor/operations.ts (hunk 5 above)
Comment (t3chguy) [reply src/editor/operations.ts:189]: Yes but this also checks that it doesn't start & end with backticks, which `hasBacktick` doesn't really imply at all

File: src/editor/operations.ts (hunk 7 above)
Comment (t3chguy) [reply src/editor/operations.ts:223]: I simplified it locally to 

image

Maybe take some inspiration from this?

Comment (yaya-usman) [reply src/editor/operations.ts:223]: Alright thanks. I'll do that ASAP. Also suddenly some tests are failing now when i push here but it all works fine on my local machine when i run `yarn test`.

File: test/editor/operations-test.js [hunk 9]
```diff
@@ -44,6 +46,74 @@ describe('editor/operations: formatting operations', () => {
...
+                expect(model.serializeParts()).toEqual([{ "text": "hell```o`w`o``rld", "type": "plain" }]);
+                formatRangeAsCode(range);
+                expect(model.serializeParts()).toEqual([{ "text": "````hell```o`w`o``rld````", "type": "plain" }]);
+            });
+
+            it('escapes backticks in between text and untoggles already formated code text', () => {
+                const renderer = createRenderer();
+                const pc = createPartCreator();
+                const model = new EditorModel([
+                    pc.plain("`hello`world`"),
```
Comment (t3chguy) [inline test/editor/operations-test.js:104]: This isn't valid so not great to test, can you update to
```
``hello`world``
```

File: src/editor/operations.ts [hunk 10]
```diff
@@ -205,7 +206,9 @@ export function formatRangeAsCode(range: Range): void {
             parts.push(partCreator.newline());
         }
     } else {
-        toggleInlineFormat(range, "`");
+        const fenceLen = longestBacktickSequence(range.text);
+        const hasInlineFormatting = range.text.startsWith("`") && range.text.endsWith("`");
+        toggleInlineFormat(range, "`".repeat(hasInlineFormatting ? fenceLen : fenceLen + 1));
```
Comment (t3chguy) [inline src/editor/operations.ts:211]: Could we get a comment here to explain the ternary please

**Review by t3chguy — COMMENTED** [review]
You also have some merge conflicts to resolve

File: test/editor/operations-test.js (hunk 9 above)
Comment (yaya-usman) [reply test/editor/operations-test.js:104]: I think i'll just remove it entirely, because i have already written your suggested update. probably you missed it.

File: src/editor/operations.ts (hunk 10 above)
Comment (yaya-usman) [reply src/editor/operations.ts:211]: No problem boss

**Comment by yaya-usman** [comment]
> You also have some merge conflicts to resolve

I don't know which lines are causing the conflict.

**Comment by t3chguy** [comment]
> I don't know which lines are causing the conflict.

That's not how merge conflicts work, you should run `git pull origin develop` and follow the interactive conflict resolution to figure out what the right output combination is. Or you can rebase but that'd then require force pushing so is not ideal.

**Comment by yaya-usman** [comment]
> > I don't know which lines are causing the conflict.
> 
> That's not how merge conflicts work, you should run `git pull origin develop` and follow the interactive conflict resolution to figure out what the right output combination is. Or you can rebase but that'd then require force pushing so is not ideal.

I did that, no conflict whatsoever. Also tests suddenly start failing when i push here but all necessary test passed perfectly on my PC. Which is why i said i don't really understand the problem now.

**Comment by yaya-usman** [comment]
> > > I don't know which lines are causing the conflict.
> > 
> > 
> > That's not how merge conflicts work, you should run `git pull origin develop` and follow the interactive conflict resolution to figure out what the right output combination is. Or you can rebase but that'd then require force pushing so is not ideal.
> 
> I did that, no conflict whatsoever. Also tests suddenly start failing when i push here but all necessary test passed perfectly on my PC. Which is why i said i don't really understand the problem now.

I have figured why it's failing and having conflicts, i didn't fetch upstream so my forked repo is not in sync or updated with the main repo

File: test/editor/operations-test.js (hunk 9 above)
Comment (t3chguy) [reply test/editor/operations-test.js]: This was the only test you had for the toggling behaviour though, we need a test which asserts that undoing the code fence also works as expected for larger (>1) fences

Comment (yaya-usman) [reply test/editor/operations-test.js]: so basically you're saying i should write a test for something like this: from ```hello``world``` to hello``world right?

Comment (t3chguy) [reply test/editor/operations-test.js]: Yup! To make sure it doesn't regress in the future

**Comment by yaya-usman** [comment]
@t3chguy hello, i have done as instructed but somehow the e2e test is failing, also should i squash all my commits into 1 and push to avoid unnecessary commits , what do you think?

**Comment by yaya-usman** [comment]
@t3chguy, can you pls check this out.

**Review by t3chguy — APPROVED** [review]
Great, thanks for persevering with this!!

**Comment by yaya-usman** [comment]
> Great, thanks for persevering with this!!

Thank you too, i learnt a lot.
